common/src/main/java/io/netty/util/concurrent/FastThreadLocalThread.java

```diff
@@ -96,19 +106,108 @@ public final void setThreadLocalMap(InternalThreadLocalMap threadLocalMap) {
      */
+    @Deprecated
     public static boolean willCleanupFastThreadLocals(Thread thread) {
         return thread instanceof FastThreadLocalThread &&
                 ((FastThreadLocalThread) thread).willCleanupFastThreadLocals();
     }
 
+    /**
+     * Returns {@code true} if {@link FastThreadLocal#removeAll()} will be called once {@link Thread#run()} completes.
+     */
+    public static boolean currentThreadWillCleanupFastThreadLocals() {
+        // intentionally doesn't accept a thread parameter to work with ScopedValue in the future
+        Thread currentThread = currentThread();
+        if (currentThread instanceof FastThreadLocalThread) {
+            return ((FastThreadLocalThread) currentThread).willCleanupFastThreadLocals();
+        }
+        return isFastThreadLocalVirtualThread();
+    }
+
+    /**
+     * Returns {@code true} if this thread supports {@link FastThreadLocal}.
+     */
+    public static boolean currentThreadHasFastThreadLocal() {
+        // intentionally doesn't accept a thread parameter to work with ScopedValue in the future
+        return currentThread() instanceof FastThreadLocalThread || isFastThreadLocalVirtualThread();
+    }
+
+    private static boolean isFastThreadLocalVirtualThread() {
+        long[] arr = fallbackThreads.get();
+        if (arr == null) {
+            return false;
+        }
+        return Arrays.binarySearch(arr, Thread.currentThread().getId()) >= 0;
+    }
+
+    /**
+     * Run the given task with {@link FastThreadLocal} support. This call should wrap the runnable for any thread that
+     * is long-running enough to make treating it as a {@link FastThreadLocalThread} reasonable, but that can't
+     * actually extend this class (e.g. because it's a virtual thread). Netty will use optimizations for recyclers and
+     * allocators as if this was a {@link FastThreadLocalThread}.
+     * <p>This method will clean up any {@link FastThreadLocal}s at the end, and
+     * {@link #currentThreadWillCleanupFastThreadLocals()} will return {@code true}.
+     * <p>At the moment, {@link FastThreadLocal} uses normal {@link ThreadLocal} as the backing storage here, but in
+     * the future this may be replaced with scoped values, if semantics can be preserved and performance is good.
+     *
+     * @param runnable The task to run
+     */
+    public static void runWithFastThreadLocal(Runnable runnable) {
+        long id = currentThread().getId();
+        if (currentThread() instanceof FastThreadLocalThread) {
+            throw new IllegalStateException("Caller is a real FastThreadLocalThread");
+        }
+        fallbackThreads.updateAndGet(arr -> {
+            if (arr == null) {
+                return new long[] { id };
+            }
+            int index = Arrays.binarySearch(arr, id);
+            if (index >= 0) {
+                throw new IllegalStateException("Reentrant call to run()");
+            }
+            index = ~index;
```

Comment:
the javadoc states it returns 
```
-(insertion point) - 1
```
so, please add a line as
```
// same as -(index + 1)
```

Reply:
to be honest, i always remember the `~index` shortcut, but not the `-(index + 1)` from the javadoc :D

Reply:
I know but IDK how many folks know about two complement by heart (I know is strange)